rust/migration/src/m20220101_000002_create_table.rs

```diff
@@ -36,14 +36,36 @@ impl MigrationTrait for Migration {
                     .col(ColumnDef::new(Job::Stdin).string().not_null())
                     .col(ColumnDef::new(Job::IsAtty).boolean().not_null())
                     .col(ColumnDef::new(Job::HiddenInfo).ezblob())
-                    .col(ColumnDef::new(Job::Stdout).ezblob())
-                    .col(ColumnDef::new(Job::Stderr).ezblob())
+                    .col(ColumnDef::new(Job::StdoutId).integer().not_null())
+                    .col(ColumnDef::new(Job::StderrId).integer().not_null())
                     .col(ColumnDef::new(Job::Status).integer().not_null())
                     .col(ColumnDef::new(Job::Runtime).double().not_null())
                     .col(ColumnDef::new(Job::Cputime).double().not_null())
                     .col(ColumnDef::new(Job::Memory).big_unsigned().not_null())
                     .col(ColumnDef::new(Job::IBytes).big_unsigned().not_null())
                     .col(ColumnDef::new(Job::OBytes).big_unsigned().not_null())
+                    .foreign_key(
+                        ForeignKeyCreateStatement::new()
+                            .name("fk-stdout_id-blob")
+                            .from_tbl(Job::Table)
+                            .from_col(Job::StdoutId)
+                            .to_tbl(Blob::Table)
+                            .to_col(Blob::Id)
+                            // A blob may not be deleted if there exists a
+                            // job that depends on it.
+                            .on_delete(ForeignKeyAction::Restrict),
```

Comment:
These in particular seem like they might be too strong? Sometimes a blob will go missing from storage and if that happens we want to delete the blob from the database, but also delete any jobs that depend on it. I think deleting stores is often not needed so I'm less concerned with that but deleting blobs seems to be more important and common.

Reply:
If a blob goes missing from storage you should just delete the whole job. Then you can either delete the blob manually or let blob eviction clean up the entry

Reply:
In fact, I claim you should _always_ delete the job first. That way you stop the bleeding with clients. Deleting the blob first makes the bleeding worse for a short amount of time

Reply:
I agree you should delete the job but there may be many jobs you need to delete not just one

Reply:
If this is a case we see happening a lot we should write code to handle the cleanup (we can use rsc_tool or the http api to delete all jobs that depend on blob x) instead of having the database work against the guarantees we want

Reply:
I cascade wouldn't be as bad if we have some way to ensure we cleanup on the remote stores with a blob is deleted. I'm not sure how we'd do that but if you have any ideas I'm open to hearing them

Reply:
The way I've been doing it in the local shared cache is that you scan both ways because you shouldn't trust either side. One scan checks for non-existant blobs that are referenced in the DB, the other scan checks for blobs that exist but are orphaned. This appears to be a required step because you can't atomically delete the blob on one side and the blob in the database. One must be deleted before the other so one of the scan directions is required. In practice I've found that both are helpful.

Reply:
Side note: We have been using on delete cascade, if you want to change that lets change it everywhere you think it makes sense while we're changing the DB.

I'm fine with it either way, using cascade only makes the delete *Easier* it doesn't actully buy us any kind of assurance so I'm quite happy to keep restrict over cascade. But I think the assurances it buys us are week (but certainly greater than the absence of assurance we get from cascade).

I will warn against thinking that this protects you from the blob store and the database falling out of sync, I do not think it accomplishes that. Distributed systems are hard.

Reply:
Yeah ok, I think I've landed on liking restrict with the following reasoning:

1) It corrects for some mistakes
2) We still need at least one direction of out of sync checks if not both but that's ok (its not a + to either side)
3) We can always do the cascade deletes manually if we need to